File: src/main/python/day25.py
def fromSnafu(c):
	if c == '2': 
	  return 2
	elif c == '1':
		 return 1
	elif c == '0':
		 return 0
	elif c == '-':
		 return -1
	else:
		 return -2

def maxSnafu(power):
	if power < 1:
		return 0
	if power == 1:
		return 2
	return 2 * power + maxSnafu(power / 5)
	
def maxPower(i):
	cur = 1
	total = 0
	while (total < i):
		cur *= 5
		total += 2 * cur
	return cur

def toSnafu(i, power):
	if power < 1:
		return ""
	maxi = maxSnafu(power / 5)
	if i > 0 and i > power + maxi:
		return '2' + toSnafu(i - 2 * power, power / 5)
	if i > 0 and i > maxi:
		return '1' + toSnafu(i - power, power / 5)	
	if -i > power + maxi:
		return '=' + toSnafu(i + 2 * power, power / 5)
	if -i > maxi:
		return '-' + toSnafu(i + power, power / 5)	
	return '0' + toSnafu(i, power / 5)
		
f = open("/storage/emulated/0/Download/input.txt").read()
total = 0

for s in f.splitlines():
  power = 1
  cs = [x for x in s]
  result = 0
  for c in reversed(cs):
  	val = fromSnafu(c)
  	result += val * power
  	power *= 5
  total += result
  power = 1
  result = 0
part1 = toSnafu(total, maxPower(total))
print(f"part1={part1}")

File: src/main/kotlin/adventofcode/y2022/Day25.kt
package adventofcode.y2022 // ktlint-disable filename

import adventofcode.readFile
import java.math.BigInteger
import java.math.BigInteger.*

fun main(args: Array<String>) {
    val total = readFile("src/main/resources/y2022/day25.txt").flatMap { line ->
        line.reversed().mapIndexed { index, c ->
            5.toBigInteger().pow(index) * fromSnafu(c).toBigInteger()
        }
    }.sumOf { it }
    print("part1="+toSnafu(total, maxPower(total)))
}

private fun maxPower(i: BigInteger): BigInteger {
    var cur = ONE
    var total = ZERO
    while (total < i) {
        cur *= 5.toBigInteger()
        total += TWO * cur
    }
    return cur
}


private fun maxSnafu(power: BigInteger): BigInteger = when {
    power < ONE -> ZERO
    power == ONE -> TWO
    else -> TWO * power + maxSnafu(power / 5.toBigInteger())
}

private fun toSnafu(i: BigInteger, power: BigInteger): String {
    if (power < ONE) return ""
    val maxi = maxSnafu(power / 5.toBigInteger())
    val toSnafuFunc = { x: BigInteger -> toSnafu(x, power / 5.toBigInteger()) }
    return when {
        i > ZERO && i > power + maxi -> '2' + toSnafuFunc(i - TWO * power)
        i > ZERO && i > maxi -> '1' + toSnafuFunc(i - power)
        -i > power + maxi -> '=' + toSnafuFunc(i + TWO * power)
        -i > maxi -> '-' + toSnafuFunc(i + power)
        else -> '0' + toSnafuFunc(i)
    }
}

private fun fromSnafu(c: Char) = when (c) {
    '2' -> 2
    '1' -> 1
    '0' -> 0
    '-' -> -1
    else -> -2
}